kernel/src/table_changes/data_read.rs

```diff
@@ -0,0 +1,231 @@
+                if selection_treemap.is_empty() {
+                    // Nothing has been selected, we do not read this data file
+                    Either::Left(iter::empty())
+                } else {
+                    let added_dv = treemap_to_bools(selection_treemap)
+                        .into_iter()
+                        .map(Not::not)
+                        .collect();
+                    println!("got dv: {:?} for type {:?}", added_dv, out_type);
+                    scan_file.scan_type = out_type;
+
+                    Either::Right(iter::once(ResolvedCDFScanFile {
+                        scan_file,
+                        selection_vector: Some(added_dv),
+                    }))
+                }
+            }
+
+            // Retrieve the deletion vector from the add action and remove action
+            let add_dv = scan_file
+                .dv_info
+                .get_treemap(engine, table_root)?
+                .unwrap_or(Default::default());
+            let rm_dv = rm_dv
+                .get_treemap(engine, table_root)?
+                .unwrap_or(Default::default());
+
+            // We calculate the deletion vectors as follows. Note that logically the `rm_dv` is the
+            // beginning state of the commit, and `add_dv` is the final state of the commit. In
+            // other words the dv went from being `rm_dv` to become `add_dv`.
+            //
+            // 1. First, find the `rm_dv XOR add_dv`. This sets the bits for all rows that have
+            //    been changed in this commit.
+            // 2. Mask the set of changed bits by either the remove or add deletion vector. The
+            //    cases are as follows:
+            //    - If we mask by `rm_dv`, then the row went from 1 (deleted) to 0 (restored).
+            //      Hence this row has been added.
+            //    - If we mask by `add_dv`, then the row went fro 0 (present) to 1 (deleted). Hence
+            //      this row has been deleted.
+
+            let xor_dv = &rm_dv ^ &add_dv;
+            println!(
+                "Remove dv: {:?}, add_dv: {:?}, xor_dv: {:?}",
+                rm_dv, add_dv, xor_dv
+            );
+            let added_selection_treemap = &xor_dv & rm_dv;
+            let removed_selection_treemap = xor_dv & add_dv;
+
+            // Generate the [`ResolvedCDFScanFile`] for remove if there are rows selected
+            let removed = treemap_to_iter(
+                removed_selection_treemap,
+                scan_file.clone(),
+                CDFScanFileType::Remove,
+            );
+
+            // Generate the [`ResolvedCDFScanFile`] for add if there are rows selected
+            let added = treemap_to_iter(added_selection_treemap, scan_file, CDFScanFileType::Add);
+
+            Ok(Either::Right(added.chain(removed)))
+        }
+        (_, Some(_)) => Err(Error::generic(
+            "Remove DV should only match to an add action!",
+        )),
```

Comment:
What you're saying is instead of having `UnresolvedCdfScanFile`, we move `remove_dvs` directly into `CdfScanFile`?

I'll go ahead and implement that, and you lmk